module/zfs/dmu_object.c

```diff
@@ -40,6 +46,72 @@ dmu_object_alloc(objset_t *os, dmu_object_type_t ot, int blocksize,
 	    0, tx);
 }
 
+void dnode_allocate_structures(objset_t *os, dmu_buf_impl_t *db,
```

Comment:
it looks like this code is copied from `dnode_hold_impl()`.  Can we restructure that to call this helper?